File: src/application/use-cases/user/create-user.usecase.ts
import { Injectable } from '@nestjs/common';
import { UseCase } from '../use-case';
import { UserRepository } from '@/application/repositories/user-repository';
import { User } from '@/application/entities/user';
import * as bcrypt from 'bcrypt';

export interface CreateUserCommand {
  username: string;
  password: string;
}

export interface CreateUserResponse {
  id: string;
  username: string;
  createdAt: Date;
}

@Injectable()
export class CreateUserUseCase
  implements UseCase<CreateUserCommand, CreateUserResponse>
{
  constructor(private readonly repository: UserRepository) {}

  async execute(command: CreateUserCommand): Promise<CreateUserResponse> {
    const saltOrRounds = 10;
    const passwordHashed = await bcrypt.hash(command.password, saltOrRounds);
    const user = new User({
      ...command,
      password: passwordHashed,
      roles: ['SERVEUR'],
    });

    await this.repository.create(user);

    return {
      id: user.id,
      username: user.username,
      createdAt: user.createdAt,
    };
  }
}

File: src/main/java/com/opendevup/shop/application/usecases/user/CreateUserUseCase.java
package com.opendevup.shop.application.usecases.user;

import com.opendevup.shop.application.command.CreateUserCommand;
import com.opendevup.shop.application.gateways.repositories.UserRepository;
import com.opendevup.shop.application.gateways.services.SecurityService;
import com.opendevup.shop.application.responses.CreateUserResponse;
import com.opendevup.shop.application.usecases.UseCase;
import com.opendevup.shop.domain.User;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
@Service
public class CreateUserUseCase implements UseCase<CreateUserCommand, Mono<CreateUserResponse>> {

    private final UserRepository userRepository;
    private final SecurityService securityService;

    @Override
    public Mono<CreateUserResponse> execute(CreateUserCommand createUserCommand) {
        return userRepository.findByUsername(createUserCommand.username())
                .defaultIfEmpty(new User(null, null, null, null))
                .flatMap(user -> {
                    if (user.id() != null) {
                        return Mono.error(new RuntimeException("User is already exist"));
                    }
                    return userRepository.save(
                            new User(
                                    null,
                                    createUserCommand.username(),
                                    securityService.hashPassword(createUserCommand.password()),
                                    createUserCommand.role()
                            )
                    ).map(userSaved -> new CreateUserResponse(
                            userSaved.id(),
                            userSaved.username()
                    ));
                });
    }
}
